### PR TITLE
Implement regexp and nin filtering

### PR DESCRIPTION
**Pull request by yshing**
### Description
Implement `regexp` and `nin` filtering.

#### Related issues

- connect to https://github.com/strongloop/loopback-filters/issues/14

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

File: index.js [hunk 1]
```diff
@@ -135,6 +135,17 @@ function test(example, value) {
       return false;
     }
 
+    if (example.nin) {
+      if (!Array.isArray(example.nin)) throw TypeError('Invalid nin query, you must pass an array to nin query.')
+      if (example.nin.length === 0) return true; // not in [] should be always true;
+      for (var i = 0; i < example.nin.length; i++) {
+        if (example.nin[i] == value) {
```
Comment (b-admike) [inline index.js]: `example.nin[i] === value` here

File: test/filter.test.js [hunk 2]
```diff
@@ -43,9 +43,35 @@ describe('filter', function() {
...
+    applyFilter({where: {name: {regexp: /John/}}}, function(err, users) {
+      should.not.exist(err);
+      users.should.have.property('length', 1);
+      done();
+    });
+  });
+
+  it('should allow to find using nin', function(done) {
+    applyFilter({where: {name: {nin: ['George Harrison']}}}, function(err, users) {
+      console.log('err', err)
```
Comment (b-admike) [inline test/filter.test.js]: Let's handle the error with either `should.exist(err)` or `should.not.exist(err)` here, and subsequent check/s if it doesn't.

File: test/filter.test.js [hunk 3]
```diff
@@ -43,9 +43,35 @@ describe('filter', function() {
...
+      done();
+    });
+  });
   // input validation
   describe.skip('invalid input', function() {
     it(
+      'should throw if the regexp value is not string or regexp',
+      function(done) {
+        applyFilter({where: {name: {regexp: 123}}}, function(err, users) {
+          should.exist(err);
```
Comment (b-admike) [inline test/filter.test.js]: one more check here to verify the error message would be great.

**Review by b-admike — COMMENTED** [review]
@yshing thank you for your contribution. I have left some comments, and on top of those, could you add a few more test cases for other conditions in your conditional statements? For example, a test case for when you're passing in a string for `regexp` [here](https://github.com/strongloop/loopback-filters/pull/20/files#diff-168726dbe96b3ce427e7fedce31bb0bcR162). Another example would be when you don't pass in valid values to either the `nin` or `regexp` filters.

**Review by b-admike — CHANGES_REQUESTED** [review]
See above

File: index.js (hunk 1 above)
Comment (yshing) [reply index.js]: Nope, should keep the similar implementation as the current inq.
If need change, both inq and nin should get changed.

Comment (b-admike) [reply index.js]: I'm okay to keep it as is.

**Review by b-admike — APPROVED** [review]
LGTM, just needs a rebase, and we should be able to land.

**Comment by b-admike** [comment]
@yshing Thank you for your contribution. I have fixed some lint issues, squashed the commits, and landed this patch 🎉.

**Pull request merged**